### ISSUE TITLE
Improve Signup Flow and Github Single Sign-On Flow

### ISSUE THREAD

**Issue by Marak** [bigcompany/hook.io#265]
The current behavior when a use signs up with just an email address is that the email address is  slugified to a username. This is not ideal.

The signup process should allow both username and email at the same time.

In addition, it seems that if a user links their Github account after creating an account by email, it will create a new account if the email accounts are not the same. Instead, the Github account should link with the pre-existing session.

**Comment by Marak** [comment]
Marked as high priority.

**Comment by Marak** [comment]
Seems to also be a similar issue with accounts linked from Stripe.

**Comment by Marak** [comment]
We are no longer slugifying the user email into a username, which is good. Username and email are now required for all accounts before services can be accessed. There are now additional UX and API endpoints to facilitate this behavior.

I believe there is still an issue with Github SSO sometimes not lining up.

**Comment by zaverden** [comment]
hello @Marak 

I think signup via github is broken now.

I'm trying to signup via github (I dont have an account). I go to https://hook.io/login and and click `Log In with Github`. After auth on Github it sends me to https://hook.io/login/github/callback?code={my_code}. And it shows me an error (I see it in my browser window):
```json
{
    "status": "error",
    "errors": [
        {
            "property": "email",
            "constraint": "required",
            "expected": true,
            "actual": false,
            "value": "",
            "message": "Required value is missing"
        }
    ]
}
```
Then I go to main page https://hook.io and it redirects me to https://hook.io/email-required. It redirects me to `email-required` from any page I try to go. The site shows me `Log Out` button - it thinks I'm authorized. So, I'm on `Update Account with Email Address` form try to add my email. But `POST https://hook.io/email-required` with my email returns another error (I got this json from devtools):
```json
{
  "error": true,
  "message": "not found"
}
```

Is it possible to signup via guthub? Or it is supposed to be like sign up with email/password and then assign github acc for SSO?

**Comment by Marak** [comment]
@zaverden I think the remaining issue we have is for Github accounts which have a non-public email address associated with their accounts.

In other words, is it possible your setting in Github is for email to remain private?

If that is the case we know root cause of issue and I think I can resolve it.

FYI, we also had a bit of a service outage yesterday which involved an unplanned migrating our database. It's possible you tried to signup during that downtime.

Please let me know, thank you.

**Comment by zaverden** [comment]
@Marak yes, I believe I did not set a public email for my github account. 

Anyway, I was able to sign up by email/username/password and then have linked hook.io account with my github account.

**Comment by Marak** [comment]
Thanks for confirming the issue @zaverden 

I'll see if we can patch this so it won't happen again.

Please let us know if you run into any other issues.

**Comment by Marak** [comment]
This just came up again and I just now have been able to reproduce the issue.

Will try to have a resolution out before the end of the day.

**Comment by zaverden** [comment]
Hello @Marak 

Tried to start with new account and got same error [I described above](https://github.com/bigcompany/hook.io/issues/265#issuecomment-379144581)

I found that in both cases my profiles don't have a public email (on page https://github.com/settings/profile). And if I add a public email - process goes without any problems.

I hope this peace of info will help to locate and solve an issue
